cmd/registry/core/visitors.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2023 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package core
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/apigee/registry/pkg/connection"
+	"github.com/apigee/registry/server/registry/names"
+)
+
+type Visitor interface {
+	ProjectHandler() ProjectHandler
+	ApiHandler() ApiHandler
+	DeploymentHandler() DeploymentHandler
+	DeploymentRevisionHandler() DeploymentHandler
+	VersionHandler() VersionHandler
+	SpecHandler() SpecHandler
+	SpecRevisionHandler() SpecHandler
+	ArtifactHandler() ArtifactHandler
+}
+
+type VisitorOptions struct {
+	RegistryClient   connection.RegistryClient
+	AdminClient      connection.AdminClient
+	Pattern          string
+	Filter           string
+	AllowUnavailable bool
+}
+
+// Visit traverses a registry, applying the Visitor to each selected resource.
+func Visit(ctx context.Context, v Visitor, options VisitorOptions) error {
+	filter := options.Filter
+	name := options.Pattern
+	ac := options.AdminClient
+	rc := options.RegistryClient
+
+	// First try to match collection names.
```

Comment:
One thing I adopted in my [tree class](https://github.com/theganyo/registry/blob/main/cmd/registry/cmd/check/tree/tree.go#L41) is to dispatch in a switch. It can look like this, which I find much easier to read:

```
name := names.Parse()
switch name := root.(type) {
  case names.Project:
    return  ListProjects(ctx, ac, name, filter, v.ProjectHandler())
  case names.API:
    return  ListApis(ctx, ac, name, filter, v.ApiHandler())
  // etc
}
```

Reply:
thanks, I think this is evolving - for this step I wanted to touch as little as possible.